Silence UBSan errors, and avoid the mistakes of upstream patch attempts PR#1
and PR#2. This patch is a local copy of upstream PR#3.

--- picosat.c.orig
+++ picosat.c
@@ -2382,11 +2382,11 @@ REENTER:
 	  for (p = ps->added; p < ps->ahead; p++) 
 	    {
 	      lit = *p;
-	      if (lit->val && ps->levels)
+	      if (lit->val)
 		{
 		  litlevel = LIT2VAR (lit)->level;
 		  assert (litlevel <= ps->LEVEL);
-		  while (ps->levels + litlevel >= ps->levelshead)
+		  while (!ps->levels || (ps->levels + litlevel >= ps->levelshead))
 		    {
 		      if (ps->levelshead >= ps->eolevels)
 			ENLARGE (ps->levels, ps->levelshead, ps->eolevels);
@@ -2403,7 +2403,7 @@ REENTER:
 		      glue++;
 		    }
 		}
-	      else if (!lit->val)
+	      else
 		glue++;
 	    }
 
@@ -3430,9 +3430,9 @@ satisfied (PS * ps)
     return 0;
   assert (!ps->conflict);
   assert (bcp_queue_is_empty (ps));
-  if (ps->trail == NULL)
-    return ps->thead == NULL;
-  return ps->thead == ps->trail + ps->max_var;	/* all assigned */
+  return ps->trail
+      ? (ps->thead == ps->trail + ps->max_var) /* all assigned */
+      : (ps->max_var == 0); /* ps->trail was NULL, so avoid ptr arithmetic */
 }
 
 static void
-- 
2.25.1

